### PR TITLE
Add GitLab example

### PR DESCRIPTION
**Pull request by guyyb**
I have read the CLA Document and I hereby sign the CLA

### CODE REVIEW

**Comment by guyyb** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by guyyb** [comment]
recheck

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+image: gradle:latest
+
+before_script:
+  # this variable is used to tell gradle from which repository to download the build-info-extractor.jar, should be server-id/repo-name
+  - JFROG_CLI_EXTRACTORS_REMOTE=$ARTIFACTORY_SERVER_ID/$ARTIFCATORY_REPO
+
+  # Download JFrog CLI
+  -  curl -fL https://install-cli.jfrog.io | sh
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:8]:
```suggestion
  - curl -fL https://install-cli.jfrog.io | sh
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+image: gradle:latest
+
+before_script:
+  # this variable is used to tell gradle from which repository to download the build-info-extractor.jar, should be server-id/repo-name
+  - JFROG_CLI_EXTRACTORS_REMOTE=$ARTIFACTORY_SERVER_ID/$ARTIFCATORY_REPO
+
+  # Download JFrog CLI
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:7]:
```suggestion
  # Install JFrog CLI
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+image: gradle:latest
+
+before_script:
+  # this variable is used to tell gradle from which repository to download the build-info-extractor.jar, should be server-id/repo-name
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:4]:
```suggestion
  # This variable is used to tell Gradle from which repository to download the build-info-extractor.jar. Should be server-id/repo-name
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  - JFROG_CLI_EXTRACTORS_REMOTE=$ARTIFACTORY_SERVER_ID/$ARTIFCATORY_REPO
+
+  # Download JFrog CLI
+  -  curl -fL https://install-cli.jfrog.io | sh
+  # configure artifactory server
+  - jf config add $ARTIFACTORY_SERVER_ID --url=$ARTIFACTORY_URL --user=$ARTIFACTORY_USER --password=$ARTIFACTORY_PASS
+  - jf config show
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:11]: Why is it needed? Let's make this example as minimal as the other examples.
```suggestion
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+image: gradle:latest
+
+before_script:
+  # this variable is used to tell gradle from which repository to download the build-info-extractor.jar, should be server-id/repo-name
+  - JFROG_CLI_EXTRACTORS_REMOTE=$ARTIFACTORY_SERVER_ID/$ARTIFCATORY_REPO
+
+  # Download JFrog CLI
+  -  curl -fL https://install-cli.jfrog.io | sh
+  # configure artifactory server
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:9]:
```suggestion
  # Configure Artifactory server
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  - jf config add $ARTIFACTORY_SERVER_ID --url=$ARTIFACTORY_URL --user=$ARTIFACTORY_USER --password=$ARTIFACTORY_PASS
+  - jf config show
+
+  # configure gradle, set upload and download repos
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:13]:
```suggestion
  # Configure Gradle, set upload and download repositories
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  - jf gradlec --repo-resolve $ARTIFCATORY_REPO --repo-deploy $ARTIFCATORY_REPO
+
+gitlab-ci-gradle-build: #This job name will be used as the build info repository name in artifactory
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:16]:
```suggestion
# This job name will be used as the build info repository name in artifactory
gitlab-ci-gradle-build:
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  stage: build
+  script:
+    # upload the built artifact to artifactory
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:19]:
```suggestion
    # Build and upload the JAR to Artifactory
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    - jf gradle clean artifactoryPublish -b build.gradle --build-name=$CI_JOB_NAME --build-number=$CI_JOB_ID
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:20]:
```suggestion
    - jf gradle clean artifactoryPublish --build-name=$CI_JOB_NAME --build-number=$CI_JOB_ID
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    # Collect environment variables for the build
+    - jf rt bce $CI_JOB_NAME $CI_JOB_ID
+    # Collect VCS details from git and add them to the build
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:23]:
```suggestion
    # Collect VCS details from Git and add them to the build
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    - jf rt bag $CI_JOB_NAME $CI_JOB_ID
+    # Publish build info
+    - jf rt bp $CI_JOB_NAME $CI_JOB_ID
+    # Scan with xray
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:27]:
```suggestion
    # Scan with JFrog Xray
```

File: gitlab-examples/gitlab-pipline/README.md [hunk 12]
```diff
@@ -0,0 +1,23 @@
+# GitLab pipeline example
+This is an example of a GitLab pipeline which uses JFrog CLI to integrate with artifactory.
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:2]:
```suggestion
This is an example of a GitLab pipeline which uses JFrog CLI to integrate with Artifactory.
```

File: gitlab-examples/gitlab-pipline/README.md [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+This sample uses a gradle project, builds it, and push the artifact to artifactory, including build info, environment variables and SBOM.
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:3]:
```suggestion
This sample uses a Gradle project, builds it, and pushes the artifact to Artifactory, including build info, environment variables, and SBOM.
```

Comment (yahavi) [reply gitlab-examples/gitlab-pipline/README.md:3]: Does it publish an SBOM? How come?

File: gitlab-examples/gitlab-pipline/README.md [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+It than uses JFrog Xray to scan the build for OSS vulnerabilities.
+
+## Configuring the GitLab project
+1. create a GitLab project
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:7]:
```suggestion
1. Create a GitLab project
```

File: gitlab-examples/gitlab-pipline/README.md [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+2. import a gradle repository into the project
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:8]:
```suggestion
2. Import a Gradle repository into the project
```

File: gitlab-examples/gitlab-pipline/README.md [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+3. configure the following variables in GitLab (Settings->CI/CD->Variables):
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:9]:
```suggestion
3. Configure the following variables in GitLab (Settings->CI/CD->Variables):
```

File: gitlab-examples/gitlab-pipline/README.md [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+   1. ARTIFACTORY_SERVER_ID - a unique server ID used to reference the server configuration from within the pipeline code
+   2. ARTIFACTORY_URL - the artifactory server used
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:11]:
```suggestion
   2. ARTIFACTORY_URL - the Artifactory server used
```

File: gitlab-examples/gitlab-pipline/README.md [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+   3. ARTIFACTORY_USER - the user name used to access artifactory
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:12]:
```suggestion
   3. ARTIFACTORY_USER - the user name used to access Artifactory
```

File: gitlab-examples/gitlab-pipline/README.md [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+   4. ARTIFACTORY_PASS - a password or token used to validate the user
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:13]:
```suggestion
   4. ARTIFACTORY_PASS - a password used to validate the user
```

File: gitlab-examples/gitlab-pipline/README.md [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+   5. ARTIFCATORY_REPO - a virtual repository used to upload and download artifacts from artifactory
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:14]:
```suggestion
   5. ARTIFCATORY_REPO - a virtual repository used to upload and download artifacts from Artifactory
```

File: gitlab-examples/gitlab-pipline/README.md [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+4. Create a new pipeline in GitLab and copy the content of gitlab-ci.yml file to it
+
+*** sensitive variables like user or password should be masked in order to make sure their value is not exposed in the pipeline logs.
+
+## Running the pipeline
+By default GitLab pipelines are triggered whenever a change is commited to the project repo.
+However, you can manually trigger the pipeline by going to: CI/CD->Pipelines and press the "Run pipeline" button.
+to view the pipeline execution logs navigate to: CI/CD->Jobs, and click the required job execution line.
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:22]:
```suggestion
To view the pipeline execution logs navigate to: CI/CD->Jobs, and click the required job execution line.
```

**Review by yahavi — CHANGES_REQUESTED** [review]
Please see my inline comments and also - 
In our examples, we usually add all tools needed to run the example. I'm not sure about this one, but please consider adding a simple Gradle project to this project.

**Comment by guyyb** [comment]
Hi Yahav,
Accepted all of your comments, and added a link to the spring-petclinic project, I used to test this example

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml (hunk 9 above)
Comment (guyyb) [reply gitlab-examples/gitlab-pipline/gitlab-ci.yml:20]: Did not work without suppling the path to build.gradle

File: gitlab-examples/gitlab-pipline/README.md (hunk 13 above)
Comment (guyyb) [reply gitlab-examples/gitlab-pipline/README.md:3]: I assume it does, as I am also publishing BuildInfo, and scanning with XRay, but maybe I do not understand the difference yet. 
Please suggest rephrasing, if you think this is wrong.

**Comment by guyyb** [comment]
Sorry, missed a few comments.
Accepted all but 2, see comments added.

File: gitlab-examples/gitlab-pipline/README.md [hunk 22]
```diff
@@ -0,0 +1,23 @@
+# GitLab pipeline example
+This is an example of a GitLab pipeline which uses JFrog CLI to integrate with Artifactory.
+This sample uses a Gradle project, builds it, and pushes the artifact to Artifactory, including build info, environment variables, and SBOM.
+It than uses JFrog Xray to scan the build for OSS vulnerabilities.
+
+## Configuring the GitLab project
+1. Create a GitLab project
+2. import a gradle repository into the project (I used https://github.com/spring-projects/spring-petclinic)
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:8]:
```suggestion
2. Import a Gradle repository into the project (I used https://github.com/spring-projects/spring-petclinic)
```

File: gitlab-examples/gitlab-pipline/README.md [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+3. Configure the following variables in GitLab (Settings->CI/CD->Variables):
+   1. ARTIFACTORY_SERVER_ID - a unique server ID used to reference the server configuration from within the pipeline code
+   2. ARTIFACTORY_URL - the Artifactory server used
+   3. ARTIFACTORY_USER - the user name used to access Artifactory
+   4. ARTIFACTORY_PASS - a password used to validate the user
+   5. ARTIFCATORY_REPO - a virtual repository used to upload and download artifacts from Artifactory
+4. Create a new pipeline in GitLab and copy the content of gitlab-ci.yml file to it
+
+*** sensitive variables like user or password should be masked in order to make sure their value is not exposed in the pipeline logs.
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/README.md:17]:
```suggestion
*** Sensitive variables like user or password should be masked in order to make sure their value is not exposed in the pipeline logs.
```

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml [hunk 24]
```diff
@@ -0,0 +1,28 @@
+image: gradle:latest
+
+before_script:
+  # This variable is used to tell Gradle from which repository to download the build-info-extractor.jar. Should be server-id/repo-name
+  - JFROG_CLI_EXTRACTORS_REMOTE=$ARTIFACTORY_SERVER_ID/$ARTIFCATORY_REPO
```
Comment (yahavi) [inline gitlab-examples/gitlab-pipline/gitlab-ci.yml:5]: I think we should explain the motivation behind this environment variable and the instructions for configuring the remote repository. Actually, this is necessary only for air-gapped environments.
The following link contains all needed information: https://www.jfrog.com/confluence/display/CLI/CLI+for+JFrog+Artifactory#CLIforJFrogArtifactory-DownloadingtheMavenandGradleExtractorJARs
Maybe we should add it too.

**Review by yahavi — CHANGES_REQUESTED** [review]

File: gitlab-examples/gitlab-pipline/README.md [hunk 25]
```diff
@@ -0,0 +1,23 @@
+# GitLab pipeline example
+This is an example of a GitLab pipeline which uses JFrog CLI to integrate with Artifactory.
+This sample uses a Gradle project, builds it, and pushes the artifact to Artifactory, including build info, environment variables, and SBOM.
```
Comment (eyalbe4) [inline gitlab-examples/gitlab-pipline/README.md:3]: "...and SBOM." - which SBOM are you referring to?

Comment (guyyb) [reply gitlab-examples/gitlab-pipline/README.md:3]: I mean Xray SBOM, as I am also scanning with Xray

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml (hunk 24 above)
Comment (guyyb) [reply gitlab-examples/gitlab-pipline/gitlab-ci.yml]: Regarding your last comment about Gradle Extractor Jar, do you think we should add the explanation, or remove it from the example?
I feel it is adding unnecessary complexity to the example, and removed the focus from the main point.
I guess common use case will be using GitLab with access to the internet.
P.S.
Tested again, and it seems that -b flag was not  required for artifactoryPublish task, so I removed it.

Comment (yahavi) [reply gitlab-examples/gitlab-pipline/gitlab-ci.yml]: Removing JFROG_CLI_EXTRACTORS_REMOTE from the example sounds good to me.
If you want, you can point in the README GitLab enterprise users without Internet users to the link.

Comment (guyyb) [reply gitlab-examples/gitlab-pipline/gitlab-ci.yml]: I do not think there is a need to add the link at this point. Let's leave it as it is.

File: gitlab-examples/gitlab-pipline/gitlab-ci.yml (hunk 9 above)
Comment (guyyb) [reply gitlab-examples/gitlab-pipline/gitlab-ci.yml]: Removed the -b flag after testing it works

**Review by yahavi — APPROVED** [review]

**Pull request merged**